functorch/functorch/experimental/cond.py

```diff
@@ -0,0 +1,120 @@
+import torch
+from torch._C import DispatchKey, DispatchKeySet, ExcludeDispatchKeyGuard
+from functorch.experimental.ops import PyOperator
+from torch.utils._pytree import tree_flatten
+from torch.fx.experimental.proxy_tensor import ProxyTensor, get_isolated_graphmodule
+import torch.utils._pytree as pytree
+from torch.dispatch.dispatcher import dispatcher_singleton
+
+"""
+We're going to define a `cond` operation.
+In order to do this, we need implementations for each of the dispatch keys.
+"""
+from contextlib import contextmanager
+
+# TODO(voz): Move out somewhere else once other py dispatched ops need it
+@contextmanager
+def suspend_mode(mode):
+    torch._C._set_torch_dispatch_mode(None)
+    try:
+        yield
+    finally:
+        torch._C._set_torch_dispatch_mode(mode)
+
+
+def trace_cond(proxy_mode, func_overload, args, kwargs=None):
+    assert kwargs is None or not kwargs
+    pred, true_fn, false_fn, operands = args
+    # We only get to this step if we are (1) in tracing (2) In the right dispatch step
+    # As this op is recursively re-entrant, we need to reset the record
+    # TODO(voz): Make this feel better?
+    dispatcher_singleton.reset_dispatch_record()
+
+    if isinstance(operands, ProxyTensor):
+        operands = [operands]  # Little hack because * on a single ProxyTensor unpacks it
+    else:
+        operands = operands
+
+    true_graph = get_isolated_graphmodule(true_fn, operands, {})
+    false_graph = get_isolated_graphmodule(false_fn, operands, {})
+    true_name = "true_graph"
+    false_name = "false_graph"
+    proxy_mode.tracer.root.register_module(true_name, true_graph)
+    proxy_mode.tracer.root.register_module(false_name, false_graph)
+
+    if isinstance(operands, ProxyTensor):
+        operands = [operands]  # Prevent unwanted unpacking
+
+    args = (pred, true_graph, false_graph, operands)
+
+    def _unwrap_proxy(e):
+        return e.proxy if isinstance(e, ProxyTensor) else e
+
+    proxy_args = pytree.tree_map(_unwrap_proxy, args)
+
+    # Does this need random slug appended so as not to collide?
+    proxy_res = proxy_mode.tracer.create_proxy('call_function', func_overload, proxy_args, kwargs,
+                                            name="conditional")
```

Comment:
TODO(rzou): fill this in

Reply:
No - discussed offline. Adding in a manner.